src/core/ddsi/src/q_entity.c

```diff
@@ -828,77 +956,18 @@ dds_return_t new_participant_guid (ddsi_guid_t *ppguid, struct ddsi_domaingv *gv
-                     DDS_SEC_PROP_AUTH_LIBRARY_INIT,
-                     DDS_SEC_PROP_AUTH_LIBRARY_FINALIZE,
-                     DDS_SEC_PROP_CRYPTO_LIBRARY_PATH,
-                     DDS_SEC_PROP_CRYPTO_LIBRARY_INIT,
-                     DDS_SEC_PROP_CRYPTO_LIBRARY_FINALIZE,
-                     DDS_SEC_PROP_ACCESS_LIBRARY_PATH,
-                     DDS_SEC_PROP_ACCESS_LIBRARY_INIT,
-                     DDS_SEC_PROP_ACCESS_LIBRARY_FINALIZE};
-      GVLOGDISC ("new_participant("
-                         PGUIDFMT
-                         "): using security settings from QoS\n", PGUID(*ppguid));
-
-      /* check if all required security properties exist in qos */
-      for (size_t i = 0; i < sizeof(req) / sizeof(req[0]); i++) {
-        if (!ddsi_xqos_has_prop(&pp->plist->qos, req[i], false, true)) {
-          GVERROR ("new_participant("
-                           PGUIDFMT
-                           "): required security property %s missing in Property QoS\n", PGUID(*ppguid), req[i]);
-          ret = DDS_RETCODE_PRECONDITION_NOT_MET;
-        }
-      }
-      if (ret == DDS_RETCODE_OK) {
-        ready_to_load_security = true;
-      } else {
-        goto new_pp_err_secprop;
-      }
-    } else if (gv->config.omg_security_configuration) {
-      /* For security, configuration can be provided through the configuration.
-       * However, the specification (and the plugins) expect it to be in the QoS. */
-      GVLOGDISC ("new_participant("
-                         PGUIDFMT
-                         "): using security settings from configuration\n", PGUID(*ppguid));
-      ddsi_xqos_mergein_security_config(&pp->plist->qos, &gv->config.omg_security_configuration->cfg);
-      ready_to_load_security = true;
-    }
-
-    if( q_omg_is_security_loaded( gv->security_context ) == false ){
-      if (ready_to_load_security && q_omg_security_load(gv->security_context, &pp->plist->qos) < 0) {
-        GVERROR("Could not load security\n");
-        ret = DDS_RETCODE_NOT_ALLOWED_BY_SECURITY;
-        goto new_pp_err_secprop;
-      }
-    } else {
-      GVLOGDISC ("new_participant("
-                               PGUIDFMT
-                               "): security is already loaded for this domain\n", PGUID(*ppguid));
-    }
-
-    if (!q_omg_security_check_create_participant (pp, gv->config.domainId))
-    {
-      ret = DDS_RETCODE_NOT_ALLOWED_BY_SECURITY;
-      goto not_allowed;
-    }
-    *ppguid = pp->e.guid;
+  if ((ret = check_and_load_security_config (gv, ppguid, &pp->plist->qos)) != DDS_RETCODE_OK)
+    goto not_allowed;
+  if ((ret = q_omg_security_check_create_participant (pp, gv->config.domainId)) != DDS_RETCODE_OK)
+    goto not_allowed;
+  *ppguid = pp->e.guid;
+#else
+  if (ddsi_xqos_has_prop_prefix (&pp->plist->qos, "dds.sec."))
+  {
+    /* disallow creating a participant with a security configuration if there is support for security
+       has been left out */
+    ret = DDS_RETCODE_NOT_ALLOWED_BY_SECURITY;
```

Comment:
Maybe replace this error code by DDS_RETCODE_PRECONDITION_NOT_MET?

Reply:
I had a debate about it with myself before pushing it to GitHub, and I just don't know what is best. My reasoning was that "not allowed by security" at least makes it immediately obvious it has something to do with the security configuration, whereas "precondition not met" is one of those overly generic ones. But at the same time, it might well make one go looking in exactly the wrong place: trying to find out what is wrong with the configuration one has specified ...

I'll mull it over a bit in the hope of getting a few votes either way before having to make the call 😀

Reply:
Mulled it over in the absence of votes one way or the other. ``DDS_PRECONDITION_NOT_MET`` it will be. Thanks!